airbyte-db/db-lib/src/main/java/io/airbyte/db/instance/jobs/migrations/V0_40_18_001__AddEstimatedAndStreamStatsColumnSyncStats.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.db.instance.jobs.migrations;
+
+import org.flywaydb.core.api.migration.BaseJavaMigration;
+import org.flywaydb.core.api.migration.Context;
+import org.jooq.DSLContext;
+import org.jooq.impl.DSL;
+import org.jooq.impl.SQLDataType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * These new columns are required for progress bars.
+ * <p>
+ * - the estimated columns will contain the overall estimated records and bytes for the sync.
+ * <p>
+ * - the StreamStats column will contain a list of {@link io.airbyte.config.StreamSyncStats}. This
+ * is the estimated and currently emitted stats for each stream. As streams are varied, a JSON blob
+ * is simpler to manage for now. The alternative is a stream stats table with a foreign key to an
+ * attempt. It is not clear we will benefit from a structured relational database at this point.
+ */
+public class V0_40_18_001__AddEstimatedAndStreamStatsColumnSyncStats extends BaseJavaMigration {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(V0_40_18_001__AddEstimatedAndStreamStatsColumnSyncStats.class);
+
+  @Override
+  public void migrate(final Context context) throws Exception {
+    LOGGER.info("Running migration: {}", this.getClass().getSimpleName());
+
+    // Warning: please do not use any jOOQ generated code to write a migration.
+    // As database schema changes, the generated jOOQ code can be deprecated. So
+    // old migration may not compile if there is any generated code.
+    final DSLContext ctx = DSL.using(context.getConnection());
+    ctx.alterTable("sync_stats")
+        .add(
+            DSL.field("estimated_records", SQLDataType.BIGINT.nullable(true)),
+            DSL.field("estimated_bytes", SQLDataType.BIGINT.nullable(true)),
+            DSL.field("stream_stats", SQLDataType.JSONB.nullable(true)))
```

Comment:
Yeap. The values in the estimated records/bytes column will be a sum of the per-stream estimated record/bytes in the `stream_stats` blob.

There is still some persistence work to be done to return these in the jobs/list and jobs/get_debug_info routes. Once that is done, we will be able to start consuming them immediately.

In the context of parallel streams - these stats are per sync. So as long as the parallelisation/partitioning splits a stream cleanly across syncs, i.e. we never have half a stream in one sync and the other half in another, we will continue to be able to use this as is.

Reply:
I think having syncs run in parallel mean that we may end up having partial updates of the `stream_stats` object.
I would like to avoid having to do partial updates of a JSON value as much as possible, which means that we are putting some constraints on how we handle this in the platform. I wonder if it worth replicating the model we currently have for storing state.

Reply:
It's probably something we can keep for when we actually start digging into parallel syncs. I am not sure we need to spend more time on this at the moment.

Reply:
Ah - thanks for the context here! I think I understand what you mean. I agree we can keep it for when we tackle parallel sync and I am happy to pitch in then!

Reply:
Going to resolve this for now. Def let me know if your thoughts change!